File: Framework/Luas/Libraries/LogConfig.lua
--[[
-- log 配置类，实现获取句柄
--]]

local error = error
local os_execute = os.execute
local os_date = os.date
local io = io
local LogConfig = Class:new('LogConfig')


function LogConfig:new( config )
	if(type(config) ~= 'table') then
		error('logconfig not table')
	end
	if(empty(config['root_path'])) then
		error('config.root_path empty')
	end
	local logconfig_tmp = Class:new('LogConfig', self)
	logconfig_tmp._root_path = config['root_path']
	logconfig_tmp._file_name 			= ''
	logconfig_tmp._buffer_line_num      = 20	--内存缓冲行数。进程异常退出会丢失数据
	logconfig_tmp._suffix_date_format 	= '%Y%m%d' --格式，跟php不一致
	logconfig_tmp._is_use_buffer	    = false 		--是否启用缓冲模式,暂时写死false，后续优化支持buffer
	logconfig_tmp._new_date             = ''
	logconfig_tmp._last_date            = ''
	logconfig_tmp._last_fp              = ''
	logconfig_tmp._last_file            = ''
	--判断目录是不是存在
	if(not file_exists(logconfig_tmp._root_path)) then
		os_execute("mkdir -p " .. logconfig_tmp._root_path)
	end
	if(empty(config['file_name'])) then
		error('config.file_name empty')
	end
	logconfig_tmp._file_name = config['file_name']
	if(config['buffer_line_num']) then
		logconfig_tmp._buffer_line_num = config['buffer_line_num']
	end
	if(config['suffix_date_format']) then
		logconfig_tmp._suffix_date_format = config['suffix_date_format']
	end
	if (type(config['is_use_buffer']) == 'boolean') then	--后续在支持
		--self._is_use_buffer = config.is_use_buffer
	end
	return logconfig_tmp
end

function LogConfig:getLogFileName(  )
	self._new_date = os_date(self._suffix_date_format)
	return self._file_name .. self._new_date .. '.log'
end

function LogConfig:getFilePath(  )
	return self._root_path .. '/' .. self:getLogFileName()
end

function LogConfig:getHandle( mode )
    local file_name = self:getLogFileName()
    if( self._last_date ~= self._new_date) then
    	local file_path = self:getFilePath()
    	self:closeHandle()
    	local fp = assert(io.open(file_path, mode))
    	self._last_fp = fp
    	self._last_date = self._new_date
    	self._last_file = file_path
    end
    return self._last_fp
end

function LogConfig:closeHandle(  )
	self._last_date = ''
	if (not empty(self._last_fp)) then
		self._last_fp:close()
	end
end

return LogConfig

File: Framework/Entities/LogConfig.php
<?php
/**
 * 日志配置实体类
 *
 * ps :
 *     考虑日志类的使用情况，在判断是否重新获取文件指针的时候，并未考虑参数 mode 的变化。
 *     由于是用于日志，所以文件名是随时间变化的，也就是你两次获取的文件路径或文件指针是不同的。
 *
 * @author zhiyuan <zhiyuan12@staff.weibo.com>
 */
namespace Framework\Entities;
use Framework\Libraries\EntityBase;

class LogConfig extends EntityBase {
    private $_new_date              = '';
    private $_last_date             = '';
    private $_last_fp               = null;
    private $_last_file             = null;
    const LOCK_WAIT                 = 0.3;  //文件写锁获取等待时间，单位秒
    const BUFFER_LINE_NUM           = 20;   //内存缓冲行数。进程异常退出会丢失数据
    const IS_USE_BUFFER             = true; // 是否启用缓冲模式
    public static $DATA_STRUCT_INFO = array(
        'root_path'          => '.',
        'file_name'          => '',
        'suffix_date_format' => 'Ymd',
        'lock_wait'          => self::LOCK_WAIT,
        'buffer_line_num'    => self::BUFFER_LINE_NUM,
        'is_use_buffer'      => self::IS_USE_BUFFER
    );
    public function __construct(array $config = array()) {
        parent::__construct($config);
        if ( ! is_dir($this->root_path)) {
            $tmp = mkdir($this->root_path, 0777, true);
            if (false === $tmp) {
                //很多时候日志记录并不造成致命错误，所以不再抛出异常
                return;
            }
        }
    }
    /**
     * 获取日志文件名
     * @return string
     */
    public function getLogFileName() {
        $this->_new_date = date($this->suffix_date_format);
        $ret             = $this->file_name . $this->_new_date . '.log';
        return $ret;
    }
    /**
     * 获取日志文件绝对路径
     * @return string
     */
    public function getFilePath() {
        $file_path = $this->root_path . "/" . $this->getLogFileName();
        return $file_path;
    }
    /**
     * 获取文件指针
     * @param  string     $mode
     * @return resource
     */
    public function getHandle($mode = 'w+') {
        $file_name = $this->getLogFileName();
        if ($this->_last_date !== $this->_new_date) {
            $file_path = $this->getFilePath();
            $this->closeHandle();
            $fp               = fopen($file_path, $mode);
            $this->_last_fp   = $fp;
            $this->_last_date = $this->_new_date;
            $this->_last_file = $file_path;
        }
        return $this->_last_fp;
    }
    /**
     * 关闭文件指针
     * @return bool
     */
    public function closeHandle() {
        $this->_last_date = '';
        if (isset($this->_last_fp)) {
            fclose($this->_last_fp);
        }
    }
}